docpipe/html.py

```diff
@@ -273,17 +286,59 @@ class SplitPOnBr(Stage):
     """
 
     def __call__(self, context):
-        for br in reversed(list(context.html.xpath('.//p/br'))):
+        for br in reversed(list(context.html.xpath('.//p//br'))):
             # everything after the br moves into a new p tag
             p = context.html.makeelement('p')
-            p.text = br.tail
-
-            sibling = br.getnext()
-            while sibling is not None:
+            # reverse order of elements to be created afresh before hitting the ancestor p
+            # e.g. ['i', 'b'] in the case of <p><b><i>Text<br>text</i></b></p>
+            ancestor_tags = []
+            elements = []
+
+            # parent should be the original p, but we want to track everything on the way up too
+            parent = br.getparent()
+            while parent.tag != 'p':
+                ancestor_tags.append(parent.tag)
+                parent = parent.getparent()
+
+            for elem_tag in ancestor_tags:
+                elem = context.html.makeelement(elem_tag)
+                elements.append(elem)
+            elements.append(p)
+            # now we have e.g. [<i>, <b>, <p>]
+            # the lowest-down, <i> in the example but at minimum <p>,
+            # gets the text
+            for i, elem in enumerate(elements):
+                if i == 0:
+                    elem.text = br.tail
+
+                    # when the immediate sibling of br is a tag
+                    # e.g. <p>Text<br><i>italics</i></p>
+                    # -- make sure <italics> ends up in the right place
+                    sibling = br.getnext()
+                    while sibling is not None:
+                        elem.append(sibling)
+                        sibling = sibling.getnext()
+
+                    # when the br falls inside an intervening tag, e.g.
+                    # <p>Text 1 <b>bold 1<br>bold 2</b> text 2</p>
+                    # make sure ' text 2' ends up in the right place
+                    elem.tail = br.getparent().tail
```

Comment:
Can you add a test case for the (odd, but possible): `<p>first <br/> second</p>tail`

Reply:
I was worried for a second there 😅